File: gcp-dns-ops/assets/code-snippets/create_zone.go
// create_zone.go — Create a managed DNS zone via Go SDK
// Usage: go run ./create_zone.go
package main

import (
    "context"
    "fmt"
    "log"
    "os"
    dns "cloud.google.com/go/dns/apiv1"
    "cloud.google.com/go/dns/apiv1/dnspb"
    "google.golang.org/api/option"
    "google.golang.org/protobuf/proto"
)

func main() {
    ctx := context.Background()
    project := os.Getenv("CLOUDSDK_CORE_PROJECT")
    if project == "" { log.Fatal("CLOUDSDK_CORE_PROJECT not set") }

    client, err := dns.NewManagedZonesRESTClient(ctx,
        option.WithCredentialsFile(os.Getenv("GOOGLE_APPLICATION_CREDENTIALS")))
    if err != nil { log.Fatalf("NewManagedZonesRESTClient: %v", err) }
    defer client.Close()

    zoneName := getEnv("DNS_ZONE_NAME", "my-zone")
    dnsName := getEnv("DNS_NAME", "example.com.")
    desc := getEnv("DNS_ZONE_DESCRIPTION", "Zone created by gcp-skills")

    zone := &dnspb.ManagedZone{
        Name:        proto.String(zoneName),
        DnsName:     proto.String(dnsName),
        Description: proto.String(desc),
    }

    req := &dnspb.CreateManagedZoneRequest{
        Parent:      fmt.Sprintf("projects/%s", project),
        ManagedZone: zone,
    }

    resp, err := client.CreateManagedZone(ctx, req)
    if err != nil { log.Fatalf("CreateManagedZone: %v", err) }

    fmt.Printf("Zone created: %s\n", resp.GetName())
    fmt.Printf("Name servers: %v\n", resp.GetNameServers())
}

func getEnv(key, fallback string) string {
    if v := os.Getenv(key); v != "" { return v }
    return fallback
}

File: gcp-dns-ops/assets/code-snippets/create_zone.py
#!/usr/bin/env python3
# create_zone.py — Create a managed DNS zone via Python SDK
# Usage: python3 create_zone.py

import os

from google.cloud import dns_v1

project = os.environ["CLOUDSDK_CORE_PROJECT"]
client = dns_v1.ManagedZonesClient()

zone = dns_v1.ManagedZone(
    name=os.environ.get("DNS_ZONE_NAME", "my-zone"),
    dns_name=os.environ.get("DNS_NAME", "example.com."),
    description=os.environ.get("DNS_ZONE_DESCRIPTION", "Zone created by gcp-skills"),
    visibility=getattr(
        dns_v1.ManagedZone.Visibility,
        os.environ.get("DNS_ZONE_VISIBILITY", "PUBLIC"),
    ),
)

visibility = os.environ.get("DNS_ZONE_VISIBILITY", "PUBLIC")
if visibility == "PRIVATE":
    zone.private_visibility_config = dns_v1.ManagedZone.PrivateVisibilityConfig(
        networks=[dns_v1.ManagedZone.PrivateVisibilityConfig.Network(
            network_url=(
                "https://www.googleapis.com/compute/v1/projects/"
                f"{project}/global/networks/"
                f"{os.environ.get('DNS_VPC_NETWORK', 'default')}"
            )
        )]
    )

parent = f"projects/{project}"
created = client.create_managed_zone(parent=parent, managed_zone=zone)
print(f"Zone created: {created.name}")
print(f"Name servers: {created.name_servers}")
